Update processors README.md (#573)
open-telemetry/opentelemetry-collector: processor/README.md

The batch processor accepts spans and places them into batches grouped by node
and resource. Batching helps better compress the data and reduce the number of
outgoing connections required to transmit the data. This processor supports
both size and time based batching.

> It is highly recommended to configure the batch processor on every collector.

Reviewer: I would say that it should come before the queued_retry and after any processor that can remove data from the pipeline (eg.: the probabilistic sample processor) since we don't want the batcher to be handling data that will be dropped.

Author: ah, good call, let me address

The batch processor should be the second processor defined in the pipeline
(immediately after the memory_limiter processor).

Please refer to [config.go](batchprocessor/config.go) for the config spec.

The following configuration options can be modified:
- `num_tickers` (default = 4): Number of tickers that loop over batch buckets
- `remove_after_ticks` (default = 10): Number of ticks passed without a span arriving for a node at which time batcher is deleted

> - `send_batch_size` (default = 8192): Size after which a batch will be sent regardless of time

Reviewer: I think it is worth to mention that this is the number of spans, since it can be interpreted as size by users.

Author: Makes sense

examples on using the processor.

## <a name="memory-limiter"></a>Memory Limiter Processor

The memory_limiter processor is used to prevent out of memory situations on
the collector. Given that the amount and type of data a collector processes is
environment specific and resource utilization of the collector is also dependent
on the configured processors, it is important to put checks in place regarding
memory usage. The memory_limiter processor offers the follow safeguards:
- Ability to configure a ballast, which allocates memory and provides stability
to the heap. In summary, the ballast increases the base size of the heap so that

> GC triggers are delayed and the number of GC cycles over time is reduced.

Reviewer 2: This seems to imply the ballast is configured by memory_limiter, which is not true. The ballast is configured via "mem-ballast-size-mib" command line option. The "ballast_size_mib" configuration option in memory_limiter does not create a ballast, it merely informs memory limiter that the ballast exists, which is necessary for correct operation. In the future we will be able to get rid of this config option from memory limiter (we will need to add ability for components to know the command line options).

Author: Good call, fixed

memory usage. The memory_limiter processor offers the follow safeguards:
- Ability to define an interval when memory usage will be checked and if memory
usage exceeds a defined limit will trigger GC to reduce memory consumption.
- Ability to define an interval when memory usage will be compared against the
previous interval's value and if the delta exceeds a defined limit will trigger
GC to reduce memory consumption.

In addition, there is a command line option which can be used to define a ballast,
which allocates memory and provides stability to the heap. If defined, the ballast
increases the base size of the heap so that GC triggers are delayed and the number
of GC cycles over time is reduced. While the ballast is configured via the command
line, today the same value configured on the command line must also be defined in
the memory_limiter processor.

Note that while these configuration options can help mitigate out of memory
situations, they are not a replacement for properly sizing the collector. Be
aware that the ballast reserves memory that could otherwise be used for purposes
including serving configured queues. In addition, if the limit or spike thresholds

> are crossed, the triggered GC may result in dropped data.

Reviewer: ```suggestion
are crossed, the triggered GC may result in dropped data depending on the pipeline configuration.
```
My intent here is that in a proper configuration, the memory limiter will trigger the receiver to reply error to clients, and that should trigger a later retry so no actual data loss in that case.

Author: I added this a bit later for exactly that reason -- let me know if you think it should be repeated here.

The memory_limiter
processor should be the first processor defined in the pipeline (immediately after
the receivers). This is to ensure that backpressure can be sent to applicable
receivers and minimize the likelihood of dropped data when the memory_limiter kicks
in.

It is highly recommended to configure ballast command line option as well as the
memory_limiter processor on every collector. The memory_limiter processor should
be the first processor defined in the pipeline (immediately after the receivers).
This is to ensure that backpressure can be sent to applicable receivers and
minimize the likelihood of dropped data when the memory_limiter kicks in.

Please refer to [config.go](memorylimiter/config.go) for the config spec.

The following configuration options __must be changed__:

> - `check_interval` (default = 0s): Time between measurements of memory

Reviewer: Note that the default value causes no checks to be performed.

Reviewer: Perhaps a bug worth fixing.

Author: Noticed -- given other parameters must also be set no a huge deal. This is why I added the bolded section above.

usage. Values below 1 second are not recommended since it can result in
unnecessary CPU consumption.
- `limit_mib` (default = 0): Maximum amount of memory, in MiB, targeted to be
allocated by the process heap. Note that typically the total memory usage of
process will be about 50MiB higher than this value.
- `spike_limit_mib` (default = 0): Maximum spike expected between the
measurements of memory usage. The value must be less than `limit_mib`.

The following configuration options can also be modified:

> - `ballast_size_mib` (default = 0): Ballast size used by the proceess.

Reviewer: ```suggestion
- `ballast_size_mib` (default = 0): Ballast size used by the proceess, it must match the value of `--mem-ballast-size-mib` command-line option.
```

Author: I fixed this in my latest commit -- thanks

memory usage. The memory_limiter processor offers the follow safeguards:

- Ability to define an interval when memory usage will be checked and if memory
usage exceeds a defined limit will trigger GC to reduce memory consumption.
- Ability to define an interval when memory usage will be compared against the
previous interval's value and if the delta exceeds a defined limit will trigger
GC to reduce memory consumption.

In addition, there is a command line option (`mem-ballast-size-mib`) which can be
used to define a ballast, which allocates memory and provides stability to the
heap. If defined, the ballast increases the base size of the heap so that GC
triggers are delayed and the number of GC cycles over time is reduced. While the
ballast is configured via the command line, today the same value configured on the
command line must also be defined in the memory_limiter processor.

Note that while these configuration options can help mitigate out of memory
situations, they are not a replacement for properly sizing and configuring the
collector. For example, if the limit or spike thresholds are crossed, the triggered

> GC may result in dropped data.

Reviewer: nit: I find the suggestion below a bit more optimistic :) - btw, multiline suggestion, yay!
```suggestion
collector. For example, if the limit or spike thresholds are crossed, the collector
will return errors to all receive operations until enough memory is freed.
```

The following configuration options can also be modified:
- `ballast_size_mib` (default = 0): Must match the `mem-ballast-size-mib`
command line option.

Examples:

```yaml
processors:
  memory_limiter:
    ballast_size_mib: 2000
    check_interval: 5s
    limit_mib: 4000
    spike_limit_mib: 500
```

Refer to [config.yaml](memorylimiter/testdata/config.yaml) for detailed
examples on using the processor.

## <a name="queued-retry"></a>Queued Retry Processor

The queued_retry processor uses a bounded queue to relay batches from the receiver
or previous processor to the next processor. Received data is enqueued immediately
if the queue is not full. At the same time, the processor has one or more workers
which consume the data in the queue by sending them to the next processor or exporter.
If relaying the data to the next processor or exporter in the pipeline fails, the
processor retries after some backoff delay depending on the configuration.

Given that if the queue is full the data will be dropped, it is important to size
the queue appropriately. The queue is kept in memory today, so the larger the queue
the more memory consumed by the collector that could otherwise be used for

> other purposes. Also, since the queue is based on batches and batch sizes are

Reviewer: The queue is pay-for-play, so if there is very little actual queueing there is no memory penalty. The memory used is proportional to the actual number of items in the queue. I mean a queue of "max size" 1 or 100K will have the same footprint if there is a single batch on the queue.

environment specific, it is not as easy to understand how much memory a queue will
consume. Finally, the queue size is dependent on the deployment model of the
collector. The agent deployment model typically has a smaller queue than the collector
deployment model.

It is highly recommended to configure the queued_retry processor on every collector
as it minimizes the likelihood of data being dropped due to delays in processing or
issues exportering the data. This processor should be the last processor defined in
the pipeline because issues that require retry are typically due to exportering.

Please refer to [config.go](queuedprocessor/config.go) for the config spec.

The following configuration options can be modified:
- `backoff_delay` (default = 5s): Time interval to wait before retrying
- `num_workers` (default = 10): Number of workers that dequeue batches

> - `queue_size` (default = 5000): Maximum number of batches allowed before drop

Reviewer: ```suggestion
- `queue_size` (default = 5000): Maximum number of batches kept on memory before start dropping data
```

memory usage. The memory_limiter processor offers the follow safeguards:

- Ability to define an interval when memory usage will be checked and if memory
usage exceeds a defined limit will trigger GC to reduce memory consumption.
- Ability to define an interval when memory usage will be compared against the
previous interval's value and if the delta exceeds a defined limit will trigger
GC to reduce memory consumption.

In addition, there is a command line option (`mem-ballast-size-mib`) which can be
used to define a ballast, which allocates memory and provides stability to the
heap. If defined, the ballast increases the base size of the heap so that GC
triggers are delayed and the number of GC cycles over time is reduced. While the
ballast is configured via the command line, today the same value configured on the
command line must also be defined in the memory_limiter processor.

Note that while these configuration options can help mitigate out of memory
situations, they are not a replacement for properly sizing and configuring the
collector. Be aware that the ballast reserves memory that could otherwise be used

> for purposes including serving configured queues. In addition, if the limit or

Author: Removed

[config.go](samplingprocessor/tailsamplingprocessor/config.go) for the config spec.

The following configuration options can be modified:

> - `decision_wait` (default = 30s): Wait time since the first span before making a sampling decision

Reviewer: ```suggestion
- `decision_wait` (default = 30s): Wait time since the first span of a trace before making a sampling decision
```

Author: Good call, fixed
